### PR TITLE
DEPR: Remove int64 uint64 float64 index part 1

### PR DESCRIPTION
**Pull request by topper-123**
This PR makes progress towards removing Int64Index, UInt64Index & FloatIndex from the code base, see #42717.

In this PR I make instantiation of `Index` with numpy numeric dtypes return a `NumericIndex` with the given dtype rather than converting to the old 64 bit only numeric index types, making all numeric dtypes available for indexes.

This is just the first part of the changes planned. In follow-ups PRs I will:
* remove Int64Index, Uint64Index & Float64Index from the top namespace (done, #49670)
* Remove int64Index, Uint64Index & Float64Index from the tess (partially done in #49682 & #49678, still some work needed)
* remove Int64Index, Uint64Index & Float64Index from the code base entirely (also internally)
* Move the functionality of `NumericIndex`into the base `Index` and remove `NumericIndex`
* Update docs to reflect the above changes

## related PRs
A part of the problems I had in #49494 was a few corner case bugs. They're pulled out into seperate PRs (#49536 & #49540). They're present in this PR at present, but will be removed after those seperate PRs have been merged.

## Notable changes from this PR:
As mentioned in #49494, one notable change from the PR is that because we now have all the numeric dtypes available for indexes, indexes created from DateTimeIndex (day, year etc.) will now be in int32, where previously they were forced to int64 (because that was the only index integer dtype available).

Using int32 is the more correct way, because the underlying `DateTimeArray` returns 32bit arrays. An example of this:

```ipython
>>> import pandas as pd
>>> 
>>> x = pd.date_range(start='1/1/2018', end='1/08/2018')
>>> x.array.day
array([1, 2, 3, 4, 5, 6, 7, 8], dtype=int32)  # 32bit, current and will not be changed in this series of PRs
>>> x.day
Int64Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int64')  # before this PR
NumericIndex([1, 2, 3, 4, 5, 6, 7, 8], dtype='int32')  # after this PR
Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int32')  # after follow-up PRs
```

The above will be included in the doc changes, when I do the docs

Progress towards #42717. See also #41272

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Out of curiosity: the approach I would have taken would have been to start with changing `Index.__new__` to return NumericIndex instead of subclasses.  Would that be counter-productive?

**Comment by topper-123** [comment]
> Out of curiosity: the approach I would have taken would have been to start with changing `Index.__new__` to return NumericIndex instead of subclasses. Would that be counter-productive?

Yes, possibly. My main problem is that on my computer intp means int64, while on some machines intp means int32. Also there have been some smaller bugs that I've found working on this that slowed me down.

I try again tonight or tomorrow and I will try the `Index.__new__` approach if I'm still having problems. Thanks for the suggestion.

**Comment by topper-123** [comment]
Alright, it's all green finally! The failure in `Python Dev / actions-311-dev (macOS-latest) (pull_request)` says `ERROR: Could not find a version that satisfies the requirement setuptools==59.2.0`, so I assume unrelated to the PR).

This is quite a big PR, with lot of dtypes changes stemming from the fact that in this PR `Index` is able to return indexes with dtypes int8/int16/int32/uint8/uint16/uint32/float16/float32. There are some follow-ups after this as explained in the OP, but those should be much easier than this PR, because all the tricky dtype stuff is handled in this PR.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -6515,10 +6522,17 @@ def insert(self, loc: int, item) -> Index:
...
-        if self._typ == "numericindex":
-            # Use self._constructor instead of Index to retain NumericIndex GH#43921
-            # TODO(2.0) can use Index instead of self._constructor
-            return self._constructor._with_infer(new_values, name=self.name)
+        if not self._is_backward_compat_public_numeric_index:
+            from pandas.core.indexes.numeric import NumericIndex
+
+            if not isinstance(self, ABCRangeIndex) or not isinstance(
+                self, NumericIndex
+            ):
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:6530]: doesnt `not isinstance(self, NumericIndex)` imply `not isinstance(self, ABCRangeIndex)`?

Comment (topper-123) [reply pandas/core/indexes/base.py:6530]: Yeah, that's true. I'll fix it in the first follow-up commit.

File: pandas/core/arrays/interval.py [hunk 2]
```diff
@@ -1750,5 +1753,6 @@ def _maybe_convert_platform_interval(values) -> ArrayLike:
         values = extract_array(values, extract_numpy=True)
 
     if not hasattr(values, "dtype"):
-        return np.asarray(values)
```
Comment (jbrockmendel) [inline pandas/core/arrays/interval.py:1753]: why is this change needed?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -563,18 +564,14 @@ def _dtype_to_subclass(cls, dtype: DtypeObj):
...
-
-            return Float64Index
-        elif dtype.kind == "u":
-            from pandas.core.api import UInt64Index
-
-            return UInt64Index
-        elif dtype.kind == "i":
-            from pandas.core.api import Int64Index
+        elif (
+            is_numeric_dtype(dtype)
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:568]: can you keep the .kind checks?  their appreciably faster

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -997,18 +994,34 @@ def astype(self, dtype, copy: bool = True):
                 new_values = astype_nansafe(values, dtype=dtype, copy=copy)
 
         # pass copy=False because any copying will be done in the astype above
-        if self._is_backward_compat_public_numeric_index:
-            # this block is needed so e.g. NumericIndex[int8].astype("int32") returns
-            # NumericIndex[int32] and not Int64Index with dtype int64.
+        if not self._is_backward_compat_public_numeric_index and not isinstance(
+            self, ABCRangeIndex
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:998]: is this not redundant?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:998]: never mind, i see this check is different from the one that was redundant

**Comment by jbrockmendel** [comment]
is there no reasonable way to split this into smaller chunks?

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/indexes/base.py:569]: Ok.

File: pandas/core/arrays/interval.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/arrays/interval.py:1751]: It may look like it's not needed. Maybe it was something I added before #50130. I'll try removing it again.

**Comment by topper-123** [comment]
Updated & rebased to take into account that #49536 has been merged into main. This Pr still builds on #50195, so if someone could erge that also, this PR can stand on its own and be able to be merged (if no more comments).

File: pandas/tests/window/test_pairwise.py [hunk 5]
```diff
@@ -99,6 +99,7 @@ def test_flex_binary_frame(method, frame):
     exp = DataFrame(
         {k: getattr(frame[k].rolling(window=10), method)(frame2[k]) for k in frame}
     )
+    exp.columns = exp.columns.astype(np.int_)
```
Comment (jbrockmendel) [inline pandas/tests/window/test_pairwise.py:102]: can this be np.int64?

Comment (jbrockmendel) [reply pandas/tests/window/test_pairwise.py:102]: ditto elsewhere.  in general i dont think we use `np.int_` often

Comment (jbrockmendel) [reply pandas/tests/window/test_pairwise.py:102]: maybe np.intp?

File: pandas/tests/indexes/test_numpy_compat.py [hunk 6]
```diff
@@ -84,8 +82,18 @@ def test_numpy_ufuncs_basic(index, func):
 
         tm.assert_index_equal(result, exp)
         if type(index) is not Index or index.dtype == bool:
-            # i.e NumericIndex
-            assert isinstance(result, Float64Index)
+            assert type(result) is NumericIndex
+            if is_complex_dtype(index):
+                assert result.dtype == "complex64"
+            elif index.dtype in ["bool", "int8", "uint8"]:
+                # TODO: find out when index.dtype is float16 vs float32 (platform issue)
```
Comment (jbrockmendel) [inline pandas/tests/indexes/test_numpy_compat.py:89]: is this still relevant or was float16 disallowed?

File: pandas/tests/groupby/test_function.py [hunk 7]
```diff
@@ -786,7 +786,7 @@ def test_nlargest_and_smallest_noop(data, groups, method):
     if method == "nlargest":
         data = list(reversed(data))
     ser = Series(data, name="a")
-    result = getattr(ser.groupby(groups), method)(n=2)
+    result = getattr(ser.groupby(np.array(groups, dtype=np.int64)), method)(n=2)
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_function.py:789]: do we still need to test the case without the explicit int64?

File: pandas/tests/window/test_pairwise.py (hunk 5 above)
Comment (topper-123) [reply pandas/tests/window/test_pairwise.py]: It can be np.int64 here, I've updated the fixture. I've also updated the whole commit to use np.intp.

File: pandas/tests/indexes/test_numpy_compat.py (hunk 6 above)
Comment (topper-123) [reply pandas/tests/indexes/test_numpy_compat.py:89]: Yeah, that's not relevant anymore. I've updated.

File: pandas/tests/groupby/test_function.py (hunk 7 above)
Comment (topper-123) [reply pandas/tests/groupby/test_function.py:789]: I think that's a good idea. I've updated to test arrays of all integer dtypes.

File: pandas/tests/util/test_assert_categorical_equal.py [hunk 8]
```diff
@@ -22,8 +22,8 @@ def test_categorical_equal_order_mismatch(check_category_order):
         msg = """Categorical\\.categories are different
 
 Categorical\\.categories values are different \\(100\\.0 %\\)
-\\[left\\]:  Int64Index\\(\\[1, 2, 3, 4\\], dtype='int64'\\)
-\\[right\\]: Int64Index\\(\\[4, 3, 2, 1\\], dtype='int64'\\)"""
+\\[left\\]:  NumericIndex\\(\\[1, 2, 3, 4]\\, dtype='int64'\\)
```
Comment (jbrockmendel) [inline pandas/tests/util/test_assert_categorical_equal.py:25]: typo here?

File: pandas/tests/frame/methods/test_set_index.py [hunk 9]
```diff
@@ -159,7 +159,7 @@ def test_set_index_cast(self):
         df = DataFrame(
             {"A": [1.1, 2.2, 3.3], "B": [5.0, 6.1, 7.2]}, index=[2010, 2011, 2012]
         )
-        df2 = df.set_index(df.index.astype(np.int32))
+        df2 = df.set_index(df.index.astype(np.int64))
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_set_index.py:162]: i think this test may no longer be needed?

File: pandas/core/arrays/datetimes.py [hunk 10]
```diff
@@ -1584,7 +1584,7 @@ def isocalendar(self) -> DataFrame:
     2017-01-06    4
     2017-01-07    5
     2017-01-08    6
-    Freq: D, dtype: int64
+    Freq: D, dtype: int32
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py:1587]: could split off the datetimelike bits?

**Comment by jbrockmendel** [comment]
could split off a bunch of diff by temporarily patching `(Int64|UInt64|Float64Index).__repr__` to say "NumericIndex"?

**Comment by topper-123** [comment]
Hi @jbrockmendel ,

I've looked into the failures after changing `np.int_` to `np.intp`, and the reason for the failures is that the use of `np.intp` is not equivalent to `np.int_` on Windows **and** `np.int_` is the default numpy integer dtype, according to [this page](https://docs.scipy.org/doc/numpy-1.10.1/user/basics.types.html).

So on all platforms  `np.array([1, 2, 3])` is equivalent to `np.array([1, 2, 3], dtype=np.int_)`

To the first point, on my Mac I have:

```python
>>> import numpy as np
>>> np.intp is np.int_
True
>>> np.int_ is np.int64
True
```
So on Mac (and Linux) it doesn't matter it we use `intp` or `int_`, they're exactly identical.

OTOH, on Windows (32-bit python installation) I have:

```python
>>> np.intp is np.int_  # different from Mac and Linux!
False
>>> np.int_ is np.int32
True
>>> np.intp  is np.intc
True
>>> type(np.int_(3))
int32
>>> type(np.intp(3))
intc # int32 or int64? This is weird
```

That `intp` is `intc` on Windows is very weird and caused many of the problems I had with this originally, because using `intp` passes all tests locally but fails on the CI Windows tests. My guess is that `intp` (and `intc`) is 32bit even on 64bit python installations on Windows, while `int_` always gives the same bit size as the numpy default (which is the correct bit size fort these tests).

So, in short & IMO, using `np.int_` is actually correct and using `np.intp` is wrong/will cause problems on Windows.

Can you comment, before I revert to `np.int_`?

**Comment by topper-123** [comment]
For example, if you look at the failures in [this windows test](https://github.com/pandas-dev/pandas/actions/runs/3837454817/jobs/6532782315), you see they're all 32bit failures after applying `np.intp`, where they previously passed when using `np.int_`.

**Comment by jbrockmendel** [comment]
OK, thanks for taking a look

**Comment by topper-123** [comment]
Ok thanks, I'll change it back then.

File: pandas/core/arrays/datetimes.py (hunk 10 above)
Comment (topper-123) [reply pandas/core/arrays/datetimes.py:1587]: No, this change is a consequence of allowing 32bit indexes, which wasn't possible before this PR.

File: pandas/tests/frame/methods/test_set_index.py (hunk 9 above)
Comment (topper-123) [reply pandas/tests/frame/methods/test_set_index.py:162]: Yes, I agree.

File: pandas/tests/util/test_assert_categorical_equal.py (hunk 8 above)
Comment (topper-123) [reply pandas/tests/util/test_assert_categorical_equal.py:25]: Yes, fixed.

**Comment by topper-123** [comment]
> could split off a bunch of diff by temporarily patching `(Int64|UInt64|Float64Index).__repr__` to say "NumericIndex"?

In a follow-up, if necessary? Getting this PR into main will make getting rid of those indexes a lot easier.

**Comment by jbrockmendel** [comment]
>> could split off a bunch of diff by temporarily patching (Int64|UInt64|Float64Index).__repr__ to say "NumericIndex"?

> In a follow-up, if necessary? Getting this PR into main will make getting rid of those indexes a lot easier.

This going in would make that idea unnecessary.  The idea of the suggestion was to make for a smaller diff here.

**Comment by topper-123** [comment]
Ok, I understand.

I've made #50550, which is this PR, except not including the commits from #50195. If that passes the CI in the newest iteration, I plan to rebase this PR to exclude the #50195 commits, which will help as #50195 is not simple.

Also there is #50571 which should be excluded from this PR after being merged.

IMO the repr issue is relatively uncomplicated, and I'm a bit worried that changing the repr could introduce other problems (e.g. misunderstanding what we're looking at).

**Comment by jbrockmendel** [comment]
the failing tests look like the same sqlalchemy stuff that is messing up the CI elsewhere, i.e. unrelated.  but being extra-careful is alright by me

**Comment by topper-123** [comment]
Ping.

This PR builds on #50195, as mentioned, and that has to be pulled in before this PR. #50571 has been merged in.

There is currently a rebase issue, I'll handle that after eventual comments and hopefully merge of #50195.

**Comment by jbrockmendel** [comment]
looks like interval tests are failing

**Comment by topper-123** [comment]
Updated after #50195 was merged + rebased. The remaining failure looks unrelated.

File: pandas/core/indexes/base.py [hunk 11]
```diff
@@ -986,18 +980,34 @@ def astype(self, dtype, copy: bool = True):
                 new_values = astype_array(values, dtype=dtype, copy=copy)
 
         # pass copy=False because any copying will be done in the astype above
-        if self._is_backward_compat_public_numeric_index:
-            # this block is needed so e.g. NumericIndex[int8].astype("int32") returns
-            # NumericIndex[int32] and not Int64Index with dtype int64.
+        if not self._is_backward_compat_public_numeric_index and not isinstance(
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:983]: so this chunk of code is going to be removed in a follow-up right?

File: pandas/tests/arrays/sparse/test_accessor.py [hunk 12]
```diff
@@ -212,7 +217,15 @@ def test_series_from_coo(self, dtype, dense_index):
 
         A = scipy.sparse.eye(3, format="coo", dtype=dtype)
         result = pd.Series.sparse.from_coo(A, dense_index=dense_index)
-        index = pd.MultiIndex.from_tuples([(0, 0), (1, 1), (2, 2)])
+
+        index_dtype = np.int64 if dense_index else np.int32
```
Comment (jbrockmendel) [inline pandas/tests/arrays/sparse/test_accessor.py:221]: is it clear we _want_ these to be different?  if not, can you add a TODO comment to that effect

File: pandas/tests/extension/decimal/test_decimal.py [hunk 13]
```diff
@@ -445,7 +445,8 @@ def DecimalArray__my_sum(self):
     result = df.groupby("id")["decimals"].agg(lambda x: x.values.my_sum())
     tm.assert_series_equal(result, expected, check_names=False)
     s = pd.Series(DecimalArray(data))
-    result = s.groupby(np.array([0, 0, 0, 1, 1])).agg(lambda x: x.values.my_sum())
+    grouper = np.array([0, 0, 0, 1, 1], dtype=np.int64)
```
Comment (jbrockmendel) [inline pandas/tests/extension/decimal/test_decimal.py:448]: i think i asked elsewhere: specifying this is benign, just affects result.index.dtype or something like that?

File: pandas/tests/frame/methods/test_to_csv.py [hunk 14]
```diff
@@ -388,7 +388,7 @@ def test_to_csv_dup_cols(self, nrows):
 
     @pytest.mark.slow
     def test_to_csv_empty(self):
-        df = DataFrame(index=np.arange(10))
+        df = DataFrame(index=np.arange(10, dtype=np.int64))
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_to_csv.py:391]: this is bc when we do to_csv followed by read_csv it doesnt roundtrip int32?

File: pandas/tests/groupby/aggregate/test_cython.py [hunk 15]
```diff
@@ -212,12 +212,12 @@ def test_cython_agg_empty_buckets_nanops(observed):
...
     # for these
     df = DataFrame([11, 12, 13], columns=["a"])
-    grps = range(0, 25, 5)
+    grps = np.arange(0, 25, 5, dtype=np.int_)
     # add / sum
     result = df.groupby(pd.cut(df["a"], grps), observed=observed)._cython_agg_general(
         "sum", alt=None, numeric_only=True
     )
-    intervals = pd.interval_range(0, 20, freq=5)
+    intervals = pd.IntervalIndex.from_breaks(np.arange(0, 21, 5))
```
Comment (jbrockmendel) [inline pandas/tests/groupby/aggregate/test_cython.py:220]: i thought IntervalArray converted 32bit ints to 64?

File: pandas/tests/window/test_groupby.py [hunk 16]
```diff
@@ -897,10 +897,11 @@ def test_as_index_false(self, by, expected_data):
         )
         tm.assert_frame_equal(result, expected)
 
-    def test_nan_and_zero_endpoints(self):
+    def test_nan_and_zero_endpoints(self, any_int_numpy_dtype):
         # https://github.com/twosigma/pandas/issues/53
+        dtype = np.dtype(any_int_numpy_dtype).type
```
Comment (jbrockmendel) [inline pandas/tests/window/test_groupby.py:902]: can we call this something other than dtype, since it isnt one

File: pandas/tests/groupby/test_nunique.py [hunk 17]
```diff
@@ -172,7 +172,7 @@ def test_nunique_preserves_column_level_names():
     # GH 23222
     test = DataFrame([1, 2, 2], columns=pd.Index(["A"], name="level_0"))
     result = test.groupby([0, 0, 0]).nunique()
-    expected = DataFrame([2], columns=test.columns)
+    expected = DataFrame([2], index=np.array([0], dtype=np.int_), columns=test.columns)
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_nunique.py:175]: is the np.int_ here needed?  ive been assuming that the `index=np.array([0])` ive been seeing has been there to ensure the indexes get np.int_ dtype.

File: pandas/tests/arrays/sparse/test_accessor.py (hunk 12 above)
Comment (topper-123) [reply pandas/tests/arrays/sparse/test_accessor.py:221]: I agree, I think they should be the same, but we need to decide which dtype should be used.

We have for *both* 32-bit and 64-bit systems:

```python
>>> import scipy.sparse
>>> A = scipy.sparse.eye(3, format="coo", dtype=dtype)
>>> A.row, A.col
(array([0, 1, 2], dtype=int32), array([0, 1, 2], dtype=int32))
```

I.e. scipy.sparse always uses int32 for index, also on 64-bit systems. The pandas way OTOH is to default to 64-bit dtypes, if an index dtype hasn't been specified, also on 32-bit systems.

So the question is if we should use the pandas or the scipy convention for index dtype when calling `pd.Series.sparse.from_coo`.

My intuition is to follow the scipy way is this instance (i.e. use 32-bit), as `pd.Series.sparse.from_coo` is about converting an existing scipy data structure, so we should follow that data structure's conventions, where possible. What do you think?

File: pandas/core/indexes/base.py (hunk 11 above)
Comment (topper-123) [reply pandas/core/indexes/base.py:983]: Yes, It will be removed in the followup #50052.

File: pandas/tests/frame/methods/test_to_csv.py (hunk 14 above)
Comment (topper-123) [reply pandas/tests/frame/methods/test_to_csv.py:391]: Yes, read_csv can't know the dtypes of its read-in columns, so it gives dtype int64 to integer columns.

File: pandas/tests/window/test_groupby.py (hunk 16 above)
Comment (topper-123) [reply pandas/tests/window/test_groupby.py:902]: Ok, I can call it `typ`.

File: pandas/tests/groupby/test_nunique.py (hunk 17 above)
Comment (topper-123) [reply pandas/tests/groupby/test_nunique.py:175]: No, not needed anyore. This is from before #49815 was merged. I'll revert it.

**Comment by topper-123** [comment]
I'll wait for your response to the question about the `pd.Series.sparse.from_coo` question before I'll push an update for the PR.

File: pandas/tests/reshape/test_cut.py [hunk 18]
```diff
@@ -87,7 +87,9 @@ def test_bins_from_interval_index_doc_example():
     # Make sure we preserve the bins.
     ages = np.array([10, 15, 13, 12, 23, 25, 28, 59, 60])
     c = cut(ages, bins=[0, 18, 35, 70])
-    expected = IntervalIndex.from_tuples([(0, 18), (18, 35), (35, 70)])
+    expected = IntervalIndex.from_arrays(
+        left=np.array([0, 18, 35]), right=np.array([18, 35, 70])
```
Comment (jbrockmendel) [inline pandas/tests/reshape/test_cut.py:91]: not a big deal but why is this necessary?

File: pandas/tests/groupby/test_function.py [hunk 19]
```diff
@@ -768,7 +768,15 @@ def test_nsmallest():
 
 @pytest.mark.parametrize(
     "data, groups",
-    [([0, 1, 2, 3], [0, 0, 1, 1]), ([0], [0])],
+    [
+        ([0, 1, 2, 3], [0, 0, 1, 1]),
+        ([0], [0]),
+        *[
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_function.py:774]: i find this pattern difficult to read.  is there a viable alternative?

Comment (jbrockmendel) [reply pandas/tests/groupby/test_function.py:774]: maybe add a new param `@pytest.mark.parametrize("dtype", [None] + tm.ALL_INT_NUMPY_DTYPES)` and then

```
if dtype is not None:
    data = np.array(data, dtype=dtype)
    groups = np.array(groups, dtype=dtype)
```

File: pandas/tests/indexes/interval/test_constructors.py [hunk 20]
```diff
@@ -291,7 +297,27 @@ def test_left_right_dont_share_data(self):
...
+            (NumericIndex(np.arange(20, 30, 0.5), dtype=np.float64), np.float64),
+            (date_range("20180101", periods=10), "<M8[ns]"),
+            (
+                date_range("20180101", periods=10, tz="US/Eastern"),
+                "datetime64[ns, US/Eastern]",
+            ),
+            (timedelta_range("1 day", periods=10), "<m8[ns]"),
+        ]
+    )
+    def breaks_and_expected_subtype(self, request):
```
Comment (jbrockmendel) [inline pandas/tests/indexes/interval/test_constructors.py:316]: is this the same as the one above?

File: pandas/tests/indexes/interval/test_interval_range.py [hunk 21]
```diff
@@ -29,7 +29,8 @@ class TestIntervalRange:
     @pytest.mark.parametrize("freq, periods", [(1, 100), (2.5, 40), (5, 20), (25, 4)])
     def test_constructor_numeric(self, closed, name, freq, periods):
         start, end = 0, 100
-        breaks = np.arange(101, step=freq)
+        dtype = np.int64 if is_integer(freq) else np.float64
```
Comment (jbrockmendel) [inline pandas/tests/indexes/interval/test_interval_range.py:32]: same as elsewhere, isnt this done automatically now?

**Review by jbrockmendel — APPROVED** [review]
Some questions, no blockers.  LGTM cc @mroeschke

File: pandas/tests/arrays/sparse/test_accessor.py (hunk 12 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/sparse/test_accessor.py:221]: id add a TODO comment saying roughly what you said here and then punt on it

File: pandas/tests/reshape/test_cut.py (hunk 18 above)
Comment (topper-123) [reply pandas/tests/reshape/test_cut.py:91]: Yeah, not needed after #50195. I'll revert.

File: pandas/tests/groupby/test_function.py (hunk 19 above)
Comment (topper-123) [reply pandas/tests/groupby/test_function.py:774]: Yes, I can do that.

File: pandas/tests/indexes/interval/test_interval_range.py (hunk 21 above)
Comment (topper-123) [reply pandas/tests/indexes/interval/test_interval_range.py:32]: Yes, after #50195... I'll revert.

File: pandas/tests/indexes/interval/test_constructors.py (hunk 20 above)
Comment (topper-123) [reply pandas/tests/indexes/interval/test_constructors.py:316]: This can be removed entirely now. I'll update.

File: pandas/tests/apply/test_series_apply.py [hunk 22]
```diff
@@ -186,7 +186,7 @@ def test_apply_datetimetz():
     # change dtype
     # GH 14506 : Returned dtype changed from int32 to int64
```
Comment (mroeschke) [inline pandas/tests/apply/test_series_apply.py:187]: Seems like this comment is not fully true now since this was changed to int32?

Comment (topper-123) [reply pandas/tests/apply/test_series_apply.py]: I've removed the comment, same below.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome, thanks @topper-123

**Comment by jbrockmendel** [comment]
this is huge, thanks @topper-123
